src/main/java/input_feed/serial/TextToSerialR1CS.java

```diff
@@ -0,0 +1,147 @@
+import java.io.FileReader;
+
+public class TextToSerialR1CS<FieldT extends AbstractFieldElementExpanded<FieldT>> {
+    private final String filePath;
+    private final int numInputs;
+    private final int numAuxiliary;
+    private final int numConstraints;
+    private final FieldT fieldParameters;
+    private boolean negateCMatrix;
+
+    public TextToSerialR1CS(
+            final String _filePath,
+            final FieldT _fieldParameters,
+            final boolean _negateCMatrix) {
+        filePath = _filePath;
+        fieldParameters = _fieldParameters;
+        negateCMatrix = _negateCMatrix;
+
+        String[] parameters = new String[3];
+        try{
+            parameters = new BufferedReader(
+                    new FileReader(filePath + ".problem_size")).readLine().split(" ");
+        } catch (Exception e){
+            System.err.println("Error: " + e.getMessage());
+        }
+        numInputs = Integer.parseInt(parameters[0]);
+        numAuxiliary = Integer.parseInt(parameters[1]);
+        numConstraints = Integer.parseInt(parameters[2]);
+    }
+
+    public TextToSerialR1CS(final String _filePath, final FieldT _fieldParameters) {
+        this(_filePath, _fieldParameters, false);
+    }
+
+    public R1CSRelation<FieldT> loadR1CS() {
+        final R1CSConstraints<FieldT> constraints = new R1CSConstraints<>();
+
+        try{
+
+            BufferedReader brA = new BufferedReader(new FileReader(filePath + ".a"));
+            BufferedReader brB = new BufferedReader(new FileReader(filePath + ".b"));
+            BufferedReader brC = new BufferedReader(new FileReader(filePath + ".c"));
+
+            for (int currRow = 0; currRow < numConstraints; currRow++){
+                LinearCombination<FieldT> A = makeRowAt(currRow, brA, false);
+                LinearCombination<FieldT> B = makeRowAt(currRow, brB, false);
+                LinearCombination<FieldT> C = makeRowAt(currRow, brC, negateCMatrix);
+
+                constraints.add(new R1CSConstraint<>(A, B, C));
+            }
+            brA.close();
+            brB.close();
+            brC.close();
+
+        } catch (Exception e){
+            System.err.println("Error: " + e.getMessage());
+        }
+
+        return new R1CSRelation<>(constraints, numInputs, numAuxiliary);
+    }
+
+    public Tuple2<Assignment<FieldT>, Assignment<FieldT>> loadWitness() {
+
+        Assignment<FieldT> primary = new Assignment<>();
+        Assignment<FieldT> auxiliary = new Assignment<>();
+
+        try{
+            String[] constraintParameters = new BufferedReader(
+                    new FileReader(filePath + ".problem_size")).readLine().split(" ");
+
+            int numPrimary = Integer.parseInt(constraintParameters[0]);
+            int numAuxiliary = Integer.parseInt(constraintParameters[1]);
+
+            final Assignment<FieldT> fullAssignment = new Assignment<>();
+
+            BufferedReader brA = new BufferedReader(new FileReader(filePath + ".aux"));
```

Comment:
nit: I'm not s huge fan of abbreviated variable names. Assuming no auto-complete, it might save a few keystrokes to write, but makes it very hard to read (which likely happens more often than writing this variable again).

How about calling this `auxFile`?

Reply:
Actually yes. Great idea. 

On another unrelated note, you reminded me that I was also considering reverting the matrix file extensions to be `mat_a, mat_b` and `mat_c` (like the pepper outputs anyway).